src/groupeddataframe/grouping.jl

```diff
@@ -178,9 +191,135 @@ function Base.getindex(gd::GroupedDataFrame, idxs::AbstractArray)
     GroupedDataFrame(gd.parent, gd.cols, new_groups, gd.idx, new_starts, new_ends)
 end
 
+# Iterator over values of grouping columns for each group
+_groupvalues(gd::GroupedDataFrame) = (Tuple(gd.parent[gd.idx[s], gd.cols]) for s in gd.starts)
+
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the values of the grouping columns for each group. These can be used to get
+the groups using the dictionary interface (by passing the key to `getindex`
+or `get`). See [`get(gd::GroupedDataFrame, key, default)`](@ref).
+
+### Returns
+
+A vector of `NamedTuple`s with names corresponding to the grouping columns.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[2]),
+                      b = repeat([2, 1], outer=[3]),
+                      c = 1:6);
+
+julia> gd = groupby(df, :a)
+GroupedDataFrame with 3 groups based on key: a
+First Group (2 rows): a = :foo
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 1     │ 4     │
+⋮
+Last Group (2 rows): a = :baz
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 2     │ 3     │
+│ 2   │ baz    │ 1     │ 6     │
+
+julia> keys(gd)
+3-element Array{NamedTuple{(:a,),Tuple{Symbol}},1}:
+ (a = :foo,)
+ (a = :bar,)
+ (a = :baz,)
+```
+"""
+function Base.keys(gd::GroupedDataFrame)
+    colnames = Tuple(_names(gd)[gd.cols])
+    coltypes = [eltype(gd.parent[:, c]) for c in gd.cols]
+    T = NamedTuple{colnames, Tuple{coltypes...}}
+    return map(T, _groupvalues(gd))
+end
+
+# Get group by values of grouped column (plain Tuple)
+function Base.getindex(gd::GroupedDataFrame, key::Tuple)
+    for (i, v) in enumerate(_groupvalues(gd))
```

Comment:
It is OK as is, but we could do it without materializing `_groupvalues(gd)` which would be a bit less expensive (but we can leave it as is and do what I propose later as an optimization - it will be relevant only for `GroupedDataFrame` with a lot of groups).

Reply:
Here `_groupvalues()` returns a generator, so it will not allocate the entire array up front.

Reply:
Ah - sorry. Then we have an issue that generator has `eltype` `Any` so it will not be super fast. But as I have said earlier I think we can leave it as is (otherwise we could test the approach using the function barrier to speed it up, but it does not guarantee to be faster so I think it is OK to leave it as is).

Reply:
AFAICT you can just remove the single-argument `_groupvalues` method and do this here instead. That sounds clearer:
```suggestion
    for i in 1:length(gd)
        v = _groupvalues(gd, i)
```